In [9]:
### Import all necessary packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report
from sklearn.preprocessing import StandardScaler, LabelEncoder

import keras
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

### Data Exploration and Preprocessing

In [2]:
### Load the dataset
df=pd.read_csv('sonardataset.csv')

In [3]:
### Data size
df.shape

(208, 61)

In [5]:
### Data information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x_1     208 non-null    float64
 1   x_2     208 non-null    float64
 2   x_3     208 non-null    float64
 3   x_4     208 non-null    float64
 4   x_5     208 non-null    float64
 5   x_6     208 non-null    float64
 6   x_7     208 non-null    float64
 7   x_8     208 non-null    float64
 8   x_9     208 non-null    float64
 9   x_10    208 non-null    float64
 10  x_11    208 non-null    float64
 11  x_12    208 non-null    float64
 12  x_13    208 non-null    float64
 13  x_14    208 non-null    float64
 14  x_15    208 non-null    float64
 15  x_16    208 non-null    float64
 16  x_17    208 non-null    float64
 17  x_18    208 non-null    float64
 18  x_19    208 non-null    float64
 19  x_20    208 non-null    float64
 20  x_21    208 non-null    float64
 21  x_22    208 non-null    float64
 22  x_

In [6]:
### first five rows of data
df.head()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60,Y
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


In [7]:
### Statistical Summary
df.describe()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_51,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


In [8]:
### Check Duplicates
df.duplicated().sum()

np.int64(0)

In [10]:
### Check Missing values
df.isnull().sum().sum()

np.int64(0)

In [12]:
pd.unique(df['Y'])

array(['R', 'M'], dtype=object)

     2 classes: Mine (M) and Rock (R)

### Model Implementation

In [4]:
### Seperate Features  and Target
target=df['Y']
features=df.drop('Y',axis=1)

In [5]:
### Encoding
lab_en=LabelEncoder()
target=lab_en.fit_transform(target)

In [8]:
### Train_Test_Split
x_train, x_test, y_train, y_test = train_test_split(features,target,test_size=0.2,random_state=100,stratify=target)

print("Training Features:", x_train.shape)
print("Testing Features :", x_test.shape)
print("Training Target  :", y_train.shape)
print("Testing Target   :", y_test.shape)

Training Features: (166, 60)
Testing Features : (42, 60)
Training Target  : (166,)
Testing Target   : (42,)


In [10]:
### Scaling
std_sca = StandardScaler()

x_train = std_sca.fit_transform(x_train)
x_test = std_sca.transform(x_test)

* The input features were normalized using StandardScaler.
* The scaler was fitted on the training data and then applied to the test data.
* This helps the ANN converge faster and improves model performance.

In [12]:
### Basic ANN Model Building

model = Sequential()

# Input layer + Hidden layer
model.add(Dense(32, input_dim=60, activation='relu'))

# Second hidden layer
model.add(Dense(16, activation='relu'))

# Output layer
model.add(Dense(1, activation='sigmoid'))

In [13]:
### Model Compilation
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

In [14]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ dense (Dense)                        │ (None, 32)                  │           1,952 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 16)                  │             528 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_2 (Dense)                      │ (None, 1)                   │              17 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,497 (9.75 KB)

 Trainable params: 2,497 (9.75 KB)

 Non-trainable params: 0 (0.00 B)

* The ANN model was successfully constructed using TensorFlow/Keras.
* The model contains two hidden layers with 32 and 16 neurons using the ReLU activation function.
* The output layer contains 1 neuron with the Sigmoid activation function for binary classification.
* The model has a total of 2,497 trainable parameters.

In [15]:
### Model Training and validation
history = model.fit(x_train, y_train, epochs=100, batch_size=1000)

Epoch 1/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 4s 4s/step - accuracy: 0.4880 - loss: 0.7559
Epoch 2/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 136ms/step - accuracy: 0.5181 - loss: 0.7380
Epoch 3/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 137ms/step - accuracy: 0.5361 - loss: 0.7210
Epoch 4/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 152ms/step - accuracy: 0.5602 - loss: 0.7047
Epoch 5/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 158ms/step - accuracy: 0.5663 - loss: 0.6892
Epoch 6/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 134ms/step - accuracy: 0.5964 - loss: 0.6742
Epoch 7/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 140ms/step - accuracy: 0.6205 - loss: 0.6596
Epoch 8/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 144ms/step - accuracy: 0.6205 - loss: 0.6456
Epoch 9/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 131ms/step - accuracy: 0.6506 - loss: 0.6321
Epoch 10/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 149ms/step - accuracy: 0.6747 - loss: 0.6193
Epoch 11/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 199ms/step - accuracy: 0.6928 - loss: 0.6069
Epoch 12/100
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 138ms/step - accuracy: 0.

In [17]:
### Prediction
y_pred_prob = model.predict(x_test)

y_pred = (y_pred_prob > 0.5).astype(int)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 157ms/step


In [24]:
### Evaluation
print("Accuracy :", accuracy_score(y_test, y_pred))
print()
print("classification Report")
print(classification_report(y_test, y_pred))

Accuracy : 0.7619047619047619

classification Report
              precision    recall  f1-score   support

           0       0.77      0.77      0.77        22
           1       0.75      0.75      0.75        20

    accuracy                           0.76        42
   macro avg       0.76      0.76      0.76        42
weighted avg       0.76      0.76      0.76        42



* The model achieved an accuracy of 76.19% on the test data.
* The precision is 75%, indicating that 75% of the samples predicted as mines were correctly classified.
* The recall is 75%, meaning the model correctly identified 75% of the actual positive samples.
* The F1-score is 75%, showing a balanced performance between precision and recall.
* Overall, the basic ANN model provides a reasonable classification performance for distinguishing between mines and rocks.

### Hyperparameter Tuning

In [25]:
!pip install keras-tuner


   -------------------- ------------------- 1/2 [keras-tuner]
   -------------------- ------------------- 1/2 [keras-tuner]
   -------------------- ------------------- 1/2 [keras-tuner]
   -------------------- ------------------- 1/2 [keras-tuner]
   -------------------- ------------------- 1/2 [keras-tuner]
   -------------------- ------------------- 1/2 [keras-tuner]
   -------------------- ------------------- 1/2 [keras-tuner]
   ---------------------------------------- 2/2 [keras-tuner]



In [26]:
def build_model(hp):

    model = Sequential()

    # First Hidden Layer
    model.add(Dense(
        units=hp.Int('units1', min_value=16, max_value=64, step=16),
        activation=hp.Choice('activation1', ['relu', 'tanh']),
        input_shape=(60,)
    ))

    # Second Hidden Layer
    model.add(Dense(
        units=hp.Int('units2', min_value=8, max_value=32, step=8),
        activation=hp.Choice('activation2', ['relu', 'tanh'])
    ))

    # Output Layer
    model.add(Dense(1, activation='sigmoid'))

    model.compile(
        optimizer=Adam(
            learning_rate=hp.Choice(
                'learning_rate',
                values=[0.01, 0.001, 0.0001]
            )
        ),
        loss='binary_crossentropy',
        metrics=['accuracy']
    )

    return model

In [28]:
import keras_tuner as kt
tuner = kt.RandomSearch(
    hypermodel=build_model,
    objective='val_accuracy',
    max_trials=10,
    overwrite=True,
    directory='random_search',
    project_name='sonar_ann'
)

In [32]:
tuner.search(
    x_train,
    y_train,
    epochs=50,
    validation_split=0.2,
    verbose=1
)

Trial 10 Complete [00h 00m 19s]
val_accuracy: 0.8823529481887817

Best val_accuracy So Far: 0.8823529481887817
Total elapsed time: 00h 04m 28s


In [34]:
best_model = tuner.get_best_models(num_models=1)[0]

In [35]:
best_hp = tuner.get_best_hyperparameters(num_trials=1)[0]

print("Units in Hidden Layer 1:", best_hp.get('units1'))
print("Activation 1:", best_hp.get('activation1'))
print("Units in Hidden Layer 2:", best_hp.get('units2'))
print("Activation 2:", best_hp.get('activation2'))
print("Learning Rate:", best_hp.get('learning_rate'))

Units in Hidden Layer 1: 64
Activation 1: tanh
Units in Hidden Layer 2: 32
Activation 2: tanh
Learning Rate: 0.01


In [37]:
y_pred_prob = best_model.predict(x_test)
y_pred = (y_pred_prob > 0.5).astype(int)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 110ms/step


### Evaluation

In [38]:
print("Accuracy :", accuracy_score(y_test, y_pred))
print()
print("classification Report")
print(classification_report(y_test, y_pred))

Accuracy : 0.7619047619047619

classification Report
              precision    recall  f1-score   support

           0       0.83      0.68      0.75        22
           1       0.71      0.85      0.77        20

    accuracy                           0.76        42
   macro avg       0.77      0.77      0.76        42
weighted avg       0.77      0.76      0.76        42



* Random Search was used to optimize the ANN hyperparameters.
* The best model used:
* Hidden Layer 1: 64 neurons
* Activation Function 1: tanh
* Hidden Layer 2: 32 neurons
* Activation Function 2: tanh
* Learning Rate: 0.01
* The highest validation accuracy obtained during tuning was 88.24%.
* When evaluated on the test dataset, the tuned model achieved an accuracy of 76.19%.

**Comparison**

| Metric | Basic ANN | Tuned ANN |
|---------|----------:|----------:|
| Accuracy | 76.19% | 76.19% |
| Precision | 75.00% | 77.00% |
| Recall | 75.00% | 77.00% |
| F1-Score | 75.00% | 76.00% |


**Conclusion**
* Random Search successfully identified a better set of hyperparameters.
* The tuned model achieved a higher validation accuracy (88.24%) during tuning.
* However, the test accuracy remained the same (76.19%) as the basic ANN model.
* This indicates that hyperparameter tuning improved performance on the validation data but did not improve generalization on the test dataset.
* Overall, both the basic ANN and the tuned ANN performed similarly on the unseen test data.

    A basic ANN model was developed to classify sonar signals as mines or rocks. The model was trained and evaluated using accuracy, precision, recall, and F1-score. Hyperparameter tuning was performed using Random Search by varying the number of neurons, activation functions, and learning rate. The tuned model identified a better hyperparameter combination, but its test accuracy remained 76.19%, which was similar to the basic ANN model. This indicates that, for this dataset, hyperparameter tuning did not improve the model's performance on unseen test data.